### PR TITLE
Make userid optional for defectdojo

### PR DESCRIPTION
**Pull request by the-simmon**
This PR solves https://github.com/secureCodeBox/secureCodeBox/issues/783.
It depends on this [PR](https://github.com/secureCodeBox/defectdojo-client-java/pull/25) from the `defectdojo-client-java` repo. So merge the other PR before this one.

### CODE REVIEW

File: hooks/persistence-defectdojo/hook/src/main/java/io/securecodebox/persistence/strategies/VersionedEngagementsStrategy.java [hunk 1]
```diff
@@ -75,13 +77,8 @@ public List<Finding> run(Scan scan, ScanFile scanResultFile) throws Exception {
...
     } else {
-      if (this.persistenceProviderConfig.isInLowPrivilegedMode()) {
-        throw new DefectDojoPersistenceException("You need to configure the id of the DefectDojo user when running the DefectDojo Hook in low privileged mode. See: https://www.securecodebox.io/docs/hooks/defectdojo#low-privileged-mode");
-      }
-      LOG.debug("Getting DefectDojo User Id via Username from the API");
-      userId = userService.searchUnique(User.builder().username(this.config.getUsername()).build())
-        .orElseThrow(() -> new DefectDojoPersistenceException("Failed to find user with name: '" + this.config.getUsername() + "'"))
-        .getId();
+      LOG.debug("Getting DefectDojo User Id via user profile API");
+      userId = userProfileService.search().get(0).getUser().getId();
```
Comment (Weltraumschaf) [inline hooks/persistence-defectdojo/hook/src/main/java/io/securecodebox/persistence/strategies/VersionedEngagementsStrategy.java:81]: This looks like a possible IOOBE: If the collection returned by `search()` the `get(0)` will throw an `IndexOutOfBoundException.

**Review by Weltraumschaf — COMMENTED** [review]
Why is this test not needed anymore? In the subject under test you have two cases: One element as expected or empty list. Now you test neither.

**Comment by the-simmon** [comment]
> Why is this test not needed anymore? In the subject under test you have two cases: One element as expected or empty list. Now you test neither.

You are right. Fixed the unit test to use the correct service now.

**Review by Weltraumschaf — APPROVED** [review]

**Review by Weltraumschaf — APPROVED** [review]

**Comment by Weltraumschaf** [comment]
@the-simmon Java test failes.

**Comment by the-simmon** [comment]
> @the-simmon Java test failes.

yes because you have to merge the pr linked in the description first because i changed the defect-dojo-client version in build.gradle

**Comment by Weltraumschaf** [comment]
https://github.com/secureCodeBox/defectdojo-client-java/pull/25 merged

**Comment by Weltraumschaf** [comment]
@the-simmon Do we need to release the client beforehand?

**Comment by the-simmon** [comment]
> @the-simmon Do we need to release the client beforehand?

yes, good point

**Comment by rseedorff** [comment]
@the-simmon you can now use the version `0.0.40-SNAPSHOT` which contains the latest changes in the underlying defectdojo-client lib.

File: hooks/persistence-defectdojo/hook/build.gradle [hunk 2]
```diff
@@ -21,7 +21,7 @@ repositories {
 dependencies {
   implementation 'io.kubernetes:client-java:12.0.0'
 
-  implementation 'io.securecodebox:defectdojo-client:0.0.35-SNAPSHOT'
+  implementation 'io.securecodebox:defectdojo-client:0.0.38-SNAPSHOT'
```
Comment (rseedorff) [inline hooks/persistence-defectdojo/hook/build.gradle:24]:
```suggestion
  implementation 'io.securecodebox:defectdojo-client:0.0.40-SNAPSHOT'
```

**Review by rseedorff — CHANGES_REQUESTED** [review]

**Review by Weltraumschaf — APPROVED** [review]

**Pull request merged**